File: bindings/swift/examples/ScreenRecord.swift
// ScreenRecord.swift
//
// ASPIRATIONAL EXAMPLE: no `Mediaway` Swift package exists yet — no
// `mediaway-device-ffi` / `mediaway-encoder-ffi` C ABI crates have shipped
// either. This shows the target ergonomics for a future Swift binding over
// Mediaway's C ABI: a C bridging header wrapped in idiomatic Swift (classes
// with `deinit`/explicit `close()`, `throws`/`try` instead of raw C status
// codes). See ../README.md and docs/spec/c-ffi.md.
//
// Mirrors examples/screen_record.rs: open screen + microphone capture, build
// an H.264 auto encoder sized to the display's real geometry, run one small
// platform-agnostic `record` loop that polls both sources and writes
// synthetic grey NV12 placeholder frames into the encode session, then
// finish the session and write the fragmented MP4 bytes to `out_screen.mp4`.
//
// `ScreenCapture`/`Microphone` (concrete backends) and the `VideoCapture`/
// `AudioCapture` protocols they conform to are all part of the imagined
// `Mediaway` package — same as `AutoVideoEncoder`/`EncodeSession`/`VideoFrame`
// in ../EncodeToMp4.swift. Only `record(...)` and the top-level script below
// are example/app code.
//
// Run (once the real package exists):
//     swift run ScreenRecord

import Foundation
import Mediaway

// MARK: - Platform-agnostic record loop

/// Poll `video` and `audio` until `duration` elapses, writing one synthetic
/// grey NV12 placeholder frame into `session` per captured video frame, and
/// draining (but not yet doing anything with) audio frames.
///
/// Typed entirely against the `VideoCapture`/`AudioCapture` protocols — this
/// function does not know or care which concrete OS backend is behind
/// either capture. `audio` is `nil` when the caller could not open a
/// microphone; recording continues video-only in that case.
func record(
    video: VideoCapture,
    audio: AudioCapture?,
    session: EncodeSession,
    width: Int,
    height: Int,
    duration: TimeInterval
) {
    let deadline = Date().addingTimeInterval(duration)
    // Synthetic NV12 placeholder (Y=128, UV=128 → grey), same shape as
    // ../EncodeToMp4.swift: width*height Y bytes + width*height/2
    // interleaved UV bytes.
    let ySize = width * height
    let uvSize = width * height / 2
    let greyNV12 = Data(repeating: 128, count: ySize + uvSize)

    var pts: Int64 = 0
    while Date() < deadline {
        do {
            if try video.pollFrame() != nil {
                // A real backend would convert the captured pixels (often a
                // GPU-resident surface) to NV12 here; this example writes
                // the synthetic placeholder instead and releases the frame
                // back to the OS once it is no longer needed.
                video.releaseFrame()

                let frame = VideoFrame(
                    pts: pts,
                    duration: 1,
                    width: width,
                    height: height,
                    pixelFormat: .nv12,
                    data: greyNV12
                )
                try session.writeFrame(frame)
                pts += 1
            }
        } catch {
            print("screen_record: capture error (\(error))")
            break
        }

        // Audio track not wired in this example — frames are drained only,
        // matching examples/screen_record.rs.
        guard let audio else { continue }
        while (try? audio.pollFrame()) != nil {}
    }
}

// MARK: - Entry point

let fps = 30
let frameRate = Rational(1, fps)

// ── 1. Open screen capture — throws if no screen-capture backend is
//    available on this platform/OS version yet. Handle gracefully instead
//    of crashing. ───────────────────────────────────────────────────────────
let screen: ScreenCapture
do {
    screen = try ScreenCapture(display: 0, frameRate: frameRate)
} catch {
    print("screen_record: screen capture unavailable (\(error)) — platform not supported yet")
    exit(0)
}
print("screen_record: \(screen.width)x\(screen.height) display")

// ── 2. Open microphone capture — also fallible; recording continues
//    video-only if no microphone is available. ─────────────────────────────
var mic: Microphone?
do {
    mic = try Microphone(sampleRate: Rational(1, 48_000))
    print("screen_record: microphone ready")
} catch {
    print("screen_record: microphone unavailable (\(error)) — continuing without audio")
    mic = nil
}

do {
    // ── 3. Build the encoder config at the capture's real geometry, then
    //    open the auto encoder + encode session. ─────────────────────────────
    var config = VideoEncodeConfig.defaults(
        codec: .h264,
        width: screen.width,
        height: screen.height,
        frameRate: frameRate
    )
    config.bitrateBps = 8_000_000

    let encoder = try AutoVideoEncoder(config: config)
    let session = try EncodeSession(encoder: encoder)

    // ── 4. Run the shared record loop for 3 seconds. ─────────────────────────
    record(
        video: screen,
        audio: mic,
        session: session,
        width: screen.width,
        height: screen.height,
        duration: 3
    )

    // ── 5. Close both capture objects. ───────────────────────────────────────
    screen.close()
    mic?.close()

    // ── 6. Finish the encode session and write the fragmented MP4 bytes. ────
    let mp4Bytes = try session.finish()
    try mp4Bytes.write(to: URL(fileURLWithPath: "out_screen.mp4"))

    print("screen_record: -> out_screen.mp4 (\(mp4Bytes.count) bytes)")
} catch {
    screen.close()
    mic?.close()
    print("screen_record: recording failed (\(error))")
    exit(1)
}

File: bindings/go/examples/screen_record.go
// Package main is an ASPIRATIONAL example.
//
// No mediaway Go binding exists yet — this file shows the target ergonomics
// for a future package wrapping mediaway-device (screen + microphone capture)
// and mediaway-pipeline (auto encoder + encode session) via cgo over the C
// ABI (see docs/spec/c-ffi.md, Tier B). It mirrors examples/screen_record.rs:
// open a screen capture and a microphone, build the same kind of auto H.264
// encode config/session as encode_to_mp4.go, run one small platform-agnostic
// record loop that polls both capture sources and writes synthetic NV12
// frames into the session, then finish the session and write the resulting
// fragmented MP4 bytes to disk.
//
// Import paths and package names below are illustrative
// ("mediaway.dev/mediaway-go/device", "mediaway.dev/mediaway-go/pipeline").
package main

import (
	"fmt"
	"os"
	"time"

	"mediaway.dev/mediaway-go/device"
	"mediaway.dev/mediaway-go/pipeline"
)

func main() {
	const (
		displayIndex  = 0 // 0 = primary display
		fps           = 30
		micSampleRate = 48_000
		bitrateBps    = 8_000_000
		recordSeconds = 3
	)

	videoTimeBase := pipeline.Rational{Num: 1, Den: fps}

	// ── 1. Open screen capture — fallible; the OS/platform backend may not ───
	// exist yet, so report it and bail out gracefully instead of crashing.
	videoCap, err := device.OpenScreenCapture(device.NewScreenCaptureConfig(displayIndex, videoTimeBase))
	if err != nil {
		fmt.Printf("screen_record: capture unavailable (%v) — platform not supported yet\n", err)
		return
	}
	defer videoCap.Close()

	// ── 2. Open the microphone — also fallible, but losing audio is not fatal: ─
	// keep recording video-only rather than aborting the whole session.
	var audioCap device.AudioCapture // stays nil if no microphone is available
	micTimeBase := pipeline.Rational{Num: 1, Den: micSampleRate}
	mic, err := device.OpenMicrophone(device.NewMicrophoneConfig(micTimeBase))
	if err != nil {
		fmt.Printf("screen_record: mic unavailable (%v) — continuing without audio\n", err)
	} else {
		defer mic.Close()
		audioCap = mic
	}

	// ── 3. Read back the geometry the capture actually settled on ────────────
	geometry := videoCap.Geometry()
	width, height := geometry.Width, geometry.Height
	fmt.Printf("screen_record: %dx%d display, audio=%v\n", width, height, audioCap != nil)

	// ── 4. Build the encode config from that geometry, open the auto encoder ──
	// and wrap it in an encode session — same building blocks as encode_to_mp4.go.
	config := pipeline.NewAutoVideoEncodeConfig(pipeline.CodecH264, width, height, videoTimeBase)
	config.BitrateBps = bitrateBps

	encoder, err := pipeline.OpenAutoEncoder(config)
	if err != nil {
		fmt.Printf("screen_record: encoder unavailable (%v) — platform not supported yet\n", err)
		return
	}
	defer encoder.Close()

	session, err := pipeline.NewEncodeSession(encoder)
	if err != nil {
		fmt.Printf("screen_record: open encode session failed: %v\n", err)
		return
	}
	defer session.Close()

	// ── 5. The one reusable record loop — knows nothing about which OS ───────
	// backend it's talking to.
	record(videoCap, audioCap, session, width, height, recordSeconds*time.Second)

	// ── 6. Flush the encoder, finalize the muxer, and write the MP4 bytes ─────
	mp4Bytes, err := session.Finish()
	if err != nil {
		fmt.Printf("screen_record: finish failed: %v\n", err)
		return
	}

	if err := os.WriteFile("out_screen.mp4", mp4Bytes, 0o644); err != nil {
		fmt.Printf("screen_record: write out_screen.mp4 failed: %v\n", err)
		return
	}

	fmt.Printf("screen_record: %dx%d -> out_screen.mp4 (%d bytes)\n", width, height, len(mp4Bytes))
}

// record polls videoCap for frames until duration elapses, writing a
// synthetic grey NV12 placeholder frame into session for each captured video
// frame, and draining (without yet doing anything with) any frames polled
// from audioCap. audioCap may be nil if no microphone was opened.
//
// videoCap and audioCap are taken purely as the device.VideoCapture /
// device.AudioCapture interfaces, so this function compiles and runs
// identically no matter which concrete OS backend sits behind them.
func record(
	videoCap device.VideoCapture,
	audioCap device.AudioCapture,
	session *pipeline.EncodeSession,
	width, height uint32,
	duration time.Duration,
) {
	deadline := time.Now().Add(duration)

	// Grey Y=128 everywhere, U/V=128 everywhere: width*height Y bytes followed
	// by width*height/2 interleaved UV bytes — same shape as encode_to_mp4.go.
	// Stands in for the real captured pixels until GPU-frame -> NV12
	// conversion is wired up.
	nv12Len := int(width)*int(height) + int(width)*int(height)/2
	greyNV12 := make([]byte, nv12Len)
	for i := range greyNV12 {
		greyNV12[i] = 128
	}

	var pts int64
	for time.Now().Before(deadline) {
		// ── Video ────────────────────────────────────────────────────────────
		frame, err := videoCap.PollFrame()
		if err != nil {
			fmt.Printf("record: capture error (%v)\n", err)
			return
		}
		if frame != nil {
			// The frame may reference GPU-resident memory (e.g. a D3D11
			// texture or IOSurface); release it back to the OS once we're
			// done with it, before writing our placeholder in its place.
			if err := videoCap.ReleaseFrame(); err != nil {
				fmt.Printf("record: release frame failed (%v)\n", err)
			}

			err := session.WriteFrame(pipeline.VideoFrame{
				PTS:         pts,
				Duration:    1,
				Width:       width,
				Height:      height,
				PixelFormat: pipeline.PixelFormatNV12,
				Data:        greyNV12,
			})
			if err != nil {
				fmt.Printf("record: write frame failed (%v)\n", err)
				return
			}
			pts++
		}

		// ── Audio ────────────────────────────────────────────────────────────
		if audioCap != nil {
			for {
				af, err := audioCap.PollFrame()
				if err != nil || af == nil {
					break
				}
				// Drained but not yet wired to an encoder/track; a second
				// (audio) track lands with the muxer's audio path.
			}
		}
	}
}
